File: _CodeTopics/LeetCode_contest/weekly/194/after/194_2.py
class Solution(object):
    def getFolderNames(self, names):
        """
        :type names: List[str]
        :rtype: List[str]
        """
        
        length = len(names)
        dic = dict()
        res = []
        for i in range(length):
            if not dic.has_key(names[i]):
                dic[names[i]] = 1
                res.append(names[i])
                ## print "names[i] is: ",names[i]
            else:
                pos = len(names[i])
                # 这里直接从字典里记录的当前最大开始
                j = dic[names[i]]
                tempname = names[i] + '(' + str(j) + ')'
                while dic.has_key(tempname):
                    j += 1
                    suffix = '(' + str(j) + ')'
                    tempname = tempname[:pos] + suffix
                # 更新这个key对应的当前最大
                dic[names[i]] = j + 1
                dic[tempname] = 1
                res.append(tempname)
                ## print "tempname is: ",tempname
        return res
    
"""
https://leetcode-cn.com/submissions/detail/80893117/

33 / 33 个通过测试用例
状态：通过
执行用时：164 ms
内存消耗：31 MB
"""

File: _CodeTopics/LeetCode/1401-1600/001487--weekly194_2/tran/194_2.cpp
class Solution {
public:
    vector<string> getFolderNames(vector<string>& names) {

        // 第 194 场周赛第二题
        // https://leetcode-cn.com/submissions/detail/80893117/
        
        int length = names.size();
        unordered_map<string, int> dic = {};
        vector<string> res;
        for (int i = 0; i < length; ++i) {
            if (dic.find(names[i]) == dic.end()) {
                dic[names[i]] = 1;
                res.push_back(names[i]);
            } else {
                int pos = names[i].size();
                int j = dic[names[i]];
                string tmpname = names[i] + '(' + to_string(j) + ')';
                while (dic.find(tmpname) != dic.end()) {
                    j += 1;
                    string suffix = '(' + to_string(j) + ')';
                    tmpname = tmpname.substr(0, pos) + suffix;
                }
                dic[names[i]] = j + 1;
                dic[tmpname] = 1;
                res.push_back(tmpname);
            }
        }
        return res;
    }
};

/*
https://leetcode.cn/submissions/detail/408740312/

执行用时：
152 ms
, 在所有 C++ 提交中击败了
82.49%
的用户
内存消耗：
64.2 MB
, 在所有 C++ 提交中击败了
44.81%
的用户
通过测试用例：
33 / 33
*/
